workflow/controller/operator.go

```diff
@@ -808,6 +808,15 @@ func (woc *wfOperationCtx) podReconciliation() error {
 			woc.log.Warnf("pod %s deleted", nodeID)
 			woc.updated = true
 		}
+		// At this point we are certain that the pod associated with our node is running or has been run;
+		// it is safe to extract the k8s-node information given this knowledge.
+		pod, err := woc.controller.kubeclientset.CoreV1().Pods(woc.wf.ObjectMeta.Namespace).Get(nodeID, metav1.GetOptions{})
```

Comment:
We already get all the Pods once in this function, making more API calls to get the same information is unnecessary.

Perhaps change `seenPods` from `map[string]bool` to `map[string]*apiv1.Pod`? This way we can keep track of the Pods we've seen and cache their objects all at once, saving us the need to call the API here again to get information we've already seen.